### PR TITLE
[Issue #29] - Improve diagnostics of Process reading in the native code

### PR DESCRIPTION
**Pull request by oleg-nenashev**
This change should significantly improve the diagnostics info for #29 in some corner cases. All the issues below may lead to #29 or to "Failed to read PEB" (e.g. [JENKINS-8614](https://issues.jenkins-ci.org/browse/JENKINS-8614) in Jenkins tests)

Native:

- [x] Enable CRT in the Release build. It makes the DLL 5 times bigger (<10 KB => 50KB), but allows doing fine-grain logging. I can make it a non-default behavior if it is preferred
- [x] Return explicit error if the process is no longer running
- [x] Return explicit error if we try to poll the 64bit process from a 32bit JVM (probably it is a way to do it in WinP by dirty hacks, but I would not expect any other stuff in Jenkins to work reliably in such case - leads to #29 or PEB error)
- [x] Return explicit errors if the Process Memory region is not readable && we want to read EnvVars from there. For newly created processes it will lead to #29 (Header is ready, but memory region is not ready)
- [x] Return explicit error if the code is messed up, and the Environment pointer is incorrect (e.g. we read RTL_USER_PROCESS_PARAMETERS incorrectly, leads to #29)

Java:

- [x] Enable stacktraces and core dumps for the build failure
- [x] More tests

Work is in progress, because AppVeyor JVM was crashing, need to investigate. likely it is due to the `isWow64()` method, where I oversimplified the pattern once the simple approach worked.

@reviewbybees

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
So yes, I should not have removed the intelligent import code, which has been proposed for https://msdn.microsoft.com/en-us/library/ms684139.aspx . AppVeyor is not running builds as a desktop App, hence it captured the API misusage.

**Comment by oleg-nenashev** [comment]
Self-:bug:

**Comment by oleg-nenashev** [comment]
@reviewbybees @kohsuke Ready for review. I would especially appreciate feedback regarding CRT inclusion feasibility. Effectively it is required for fine-grain logging only, which I could disable for the release build (and maybe make it configurable via system property). OTOH, 40KB should not be a problem in 2017 since it needs to be loaded only once.

File: pom.xml [hunk 1]
```diff
@@ -40,6 +40,10 @@
       <plugin>
         <artifactId>maven-surefire-plugin</artifactId>
         <version>2.19.1</version>
+        <configuration>
+            <argLine>-XX:+CreateMinidumpOnCrash</argLine>
+            <trimStackTrace>false</trimStackTrace>
```
Comment (jglick) [inline pom.xml]: Sigh.

File: native/winp.h [hunk 2]
```diff
@@ -1,18 +1,26 @@
 #pragma once
 #include "stdafx.h"
 
-BOOL WINAPI KillProcessEx(IN DWORD dwProcessId, IN BOOL bTree);
-
+// Sets bit 29 in order to keep the codes in the user space
+#define reportErrorWithCode(env,code,msg)	SetLastError(code + 0x10000000); error(env,__FILE__,__LINE__,msg);
```
Comment (jglick) [inline native/winp.h]: Not using `toSystemErrorCode`?

File: src/test/java/org/jvnet/winp/PlatformSpecificProcessTest.java [hunk 3]
```diff
@@ -90,16 +90,17 @@ public void shouldNotBeCritical() throws Exception {
...
         assertFalse("The process has not been stopped yet", isAlive(p));
 
         try {
             new WinProcess(p).getCommandLine();
         } catch (WinpException ex) {
-            assertThat(ex.getMessage(), allOf(
-                    containsString("Failed to read " + getExpectedPEBName(false)), 
-                    containsString("error=299 at envvar-cmdline")));
+            assertThat(ex.getMessage(), containsString("Process with pid=" + pid + " has already stopped. Exit code is -1"));
+            assertThat(ex.getWin32ErrorCode(), equalTo(UserErrorType.PROCESS_IS_NOT_RUNNING.getSystemErrorCode()));
```
Comment (jglick) [inline src/test/java/org/jvnet/winp/PlatformSpecificProcessTest.java]: Nice.

File: native/winp.h (hunk 2 above)
Comment (oleg-nenashev) [inline native/winp.h]: Nope, because it is a native stuff. Ideally I should create enum there as well, but I firstly need to think about fatal/not-fatal errors (E.g. Process memory block unavailability is not fatal, and it is one of potential JENKINS-30782 root causes)

**Comment by oleg-nenashev** [comment]
@reviewbybees done

**Comment by oleg-nenashev** [comment]
@kohsuke said he is fine with CRT inclusion. @reviewbybees done => Merging
